src/tokenizer/sql_base.cpp

```diff
@@ -0,0 +1,225 @@
+{
+    switch (type) {
+    case sql_token_type::command:
+        os << "command";
+        break;
+    case sql_token_type::identifier:
+        os << "identifier";
+        break;
+    case sql_token_type::number:
+        os << "number";
+        break;
+    case sql_token_type::string:
+        os << "string";
+        break;
+    case sql_token_type::single_quoted_string:
+        os << "single_quoted_string";
+        break;
+    case sql_token_type::double_quoted_string:
+        os << "double_quoted_string";
+        break;
+    case sql_token_type::back_quoted_string:
+        os << "back_quoted_string";
+        break;
+    case sql_token_type::dollar_quoted_string:
+        os << "dollar_quoted_string";
+        break;
+    case sql_token_type::whitespace:
+        os << "whitespace";
+        break;
+    case sql_token_type::asterisk:
+        os << "asterisk";
+        break;
+    case sql_token_type::eol_comment:
+        os << "eol_comment";
+        break;
+    case sql_token_type::parenthesis_open:
+        os << "parenthesis_open";
+        break;
+    case sql_token_type::parenthesis_close:
+        os << "parenthesis_close";
+        break;
+    case sql_token_type::comma:
+        os << "comma";
+        break;
+    case sql_token_type::questionmark:
+        os << "questionmark";
+        break;
+    case sql_token_type::colon:
+        os << "colon";
+        break;
+    case sql_token_type::dot:
+        os << "dot";
+        break;
+    case sql_token_type::query_end:
+        os << "query_end";
+        break;
+    case sql_token_type::binary_operator:
+        os << "binary_operator";
+        break;
+    case sql_token_type::bitwise_operator:
+        os << "bitwise_operator";
+        break;
+    case sql_token_type::inline_comment:
+        os << "inline_comment";
+        break;
+    case sql_token_type::array_open:
+        os << "array_open";
+        break;
+    case sql_token_type::array_close:
+        os << "array_close";
+        break;
+    case sql_token_type::curly_brace_open:
+        os << "curly_brace_open";
+        break;
+    case sql_token_type::curly_brace_close:
+        os << "curly_brace_close";
+        break;
+    case sql_token_type::unknown:
+    default:
+        os << "unknown";
+        break;
+    }
+    return os;
+}
+
+template <typename T>
+sql_tokenizer<T>::sql_tokenizer(
+    std::string_view str, std::unordered_set<sql_token_type> skip_tokens)
+    : buffer_(str), skip_tokens_(std::move(skip_tokens))
+{
+    if (!number_regex.ok()) {
+        throw std::runtime_error("sql number regex not valid: " + number_regex.error_arg());
+    }
+}
+
+template <typename T> std::string_view sql_tokenizer<T>::extract_string(char quote)
+{
+    auto begin = index();
+    unsigned slash_count = 0;
+    while (advance()) {
+        if (peek() == '\\') {
+            // Count consecutive backslashes
+            slash_count = (slash_count + 1) % 2;
```

Comment:
- this is not how sql string escaping works:

```
glopes=> select '\';
 ?column? 
----------
 \
(1 row)

glopes=> select '''';
 ?column? 
----------
 '
(1 row)

glopes=> create table a("col""quote" text);
CREATE TABLE
````

Reply:
You're right that I need to support these escape sequences, however the escaping I implemented is supported in both mysql and postgre (conditionally).

Reply:
Fixed:
- For `pgsql`, `mysql`, `generic` I'm supporting both types of escapes.
- For sqlite, only conforming escapes, i.e.` 'str''ing''`